### PR TITLE
Track missing events in WordPress.com, Jetpack Connect and Cloud's Pricing page

### PR DESCRIPTION
**Pull request by rcanepa**
#### Changes proposed in this Pull Request

Fixes 1169247016322522-as-1196341175625977

* This PRs add missing tracks to events triggered from Calypso Plans page, Jetpack Connect Plans page, and Calypso Green Plans page (the events are basically the same between these pages).

#### Implementation notes

* Page views are tracked using the `PageViewTracker` component.
* Other events are tracked dispatching an action with `recordTracksEvent`.

#### Events

This is the list of all events (minus page views), and their names, that are being tracked:

- `calypso_cart_product_add`: dispatched when checkout adds a product to the cart (not controlled by us).
- `calypso_plans_duration_change`: dispatched when users change the duration filter (annually vs monthly).
- `calypso_plans_type_change`: dispatched when users change the product type filter (All, Security, Performance).
- `calypso_product_checkout_click`: dispatched when users click a CTA that takes them to the checkout page (the checkout page has dispatches its own event called `calypso_cart_product_add` when the product is added to the cart).
- `calypso_product_external_click`: dispatched when users click on the CRM or any other product card that redirects users outside of Calypso.
- `calypso_product_features_open`: dispatched when users open the features section of a product card.
- `calypso_product_features_close`: dispatched when users close the features section of a product card.
- `calypso_product_jpfree_click`: dispatched when users clicks on Jetpack Free CTA.
- `calypso_product_manage_click`: dispatched when users click on Manage Plan/Subscription CTA.
- `calypso_product_subtypes_click`: dispatched when users clicks on a CTA that takes them to the subtypes/options page (where they have to decide between daily or real-time versions of a product).
- `calypso_product_upsell_click`: dispatched when users clicks on a CTA that takes them to an upsell page.
- `calypso_subtypes_back_click`: dispatched when users clicks on the back button in the subtypes/options page.
- `calypso_upsell_back_click`: dispatched when users clicks on the back button in the upsell page.

#### Testing instructions

* Run this PR locally (`yarn start` and `yarn start-jetpack-cloud-p`).
* Repeat instructions for every environment (Calypso Blue, Jetpack Connect, and Calypso Green).
* Open your browser console and run `localStorage.setItem('debug', 'calypso:analytics');`.
* Follow every possible path in the Plans page making sure the events described above are being tracked. 
* Verify that a page view is tracked every time you land in a page (hard refreshes should be tracked as well).
* Verify that clicks on every CTA are being tracked (product cards, filters, navigation items, etc).

#### Screenshots

A page view being tracked in Calypso Green.
image

### CODE REVIEW

File: client/my-sites/plans-v2/details.tsx [hunk 1]
```diff
@@ -65,8 +67,19 @@ const DetailsPage = ( {
...
+	const trackUpsellView = useCallback( () => {
+		dispatch(
+			recordTracksEvent( 'calypso_product_upsell_click', {
+				site_id: siteId || undefined,
+				product_slug: productSlug,
+				duration,
+			} )
+		);
+	}, [ dispatch, duration, productSlug, siteId ] );
+
```
Comment (rcanepa) [inline client/my-sites/plans-v2/details.tsx:79]: The user is moving to the upsell page.

File: client/my-sites/plans-v2/details.tsx [hunk 2]
```diff
@@ -85,10 +98,18 @@ const DetailsPage = ( {
...
 			return;
 		}
 
+		dispatch(
+			recordTracksEvent( 'calypso_cart_product_add', {
+				site_id: siteId || undefined,
+				product_slug: slug,
+				duration,
+			} )
+		);
```
Comment (rcanepa) [inline client/my-sites/plans-v2/details.tsx:112]: The user is moving to the checkout page (a product is being added to the cart).

File: client/my-sites/plans-v2/details.tsx [hunk 3]
```diff
@@ -116,10 +137,27 @@ const DetailsPage = ( {
...
+		page( selectorPageUrl );
+	};
+
+	const viewTrackerPath = siteId
+		? `${ rootUrl }/:product/${ durationToString( duration ) }/details/:site`
+		: `${ rootUrl }/:product/${ durationToString( duration ) }/details`;
+	const viewTrackerProps = siteId
+		? { site: siteSlug, product: productSlug }
+		: { product: productSlug };
```
Comment (rcanepa) [inline client/my-sites/plans-v2/details.tsx:156]: If we have a site in the context, we want to track that as well as part of the page view tracking object.

File: client/my-sites/plans-v2/upsell.tsx [hunk 4]
```diff
@@ -203,21 +209,52 @@ const UpsellPage = ( {
...
-	const onPurchaseBothProducts = useCallback(
-		() => checkoutCb( [ productSlug, upsellProductSlug ] ),
-		[ checkoutCb, productSlug, upsellProductSlug ]
+	const onPurchaseBothProducts = useTrackCallback(
+		() => checkout( siteSlug, [ productSlug, upsellProductSlug ], urlQueryArgs ),
+		'calypso_cart_product_add',
+		{
+			site_id: siteId || undefined,
+			product_slug: productSlug,
+			upsell_product_slug: upsellProductSlug,
```
Comment (rcanepa) [inline client/my-sites/plans-v2/upsell.tsx:218]: Here we are tracking two products: the main one and the one being upsold.

**Comment by elliottprogrammer** [comment]
When I select 1 product, it appears that the `calypso_cart_product_add` event is firing twice, (but I only have one item in the cart), Scan See screenshot: 
Screenshot on 2020-09-30 at 10-08-44

When I select Backup Daily and choose the upsell to add Scan, it appears that Backup is being added to cart twice (like in my example above), but Scan is being added to cart once.  See screenshot:
Screenshot on 2020-09-30 at 10-10-34

Is this the case? Or maybe I'm interpreting the console debug events wrong?  

Can you duplicate?

**Comment by eeeeevon13** [comment]
Looks good. I thought we were also missing some tracks events on a few of the existing CTAs for each product, can you double check that each plan/product cta has a tracks event?

File: client/components/jetpack/card/jetpack-product-card/features.tsx [hunk 5]
```diff
@@ -25,20 +25,25 @@ export type Props = {
...
 	const trackShowFeatures = useTrackCallback(
 		undefined,
-		'calypso_jetpack_show_product_card_features'
+		'calypso_product_features_open',
+		trackProps
 	);
 	const trackHideFeatures = useTrackCallback(
 		undefined,
-		'calypso_jetpack_hide_product_card_features'
+		'calypso_product_features_close',
```
Comment (rcanepa) [inline client/components/jetpack/card/jetpack-product-card/features.tsx:45]: Renamed it to make it follow the same naming convention the rest of the tracks use.

File: client/components/jetpack/card/jetpack-product-card/features.tsx [hunk 6]
```diff
@@ -25,20 +25,25 @@ export type Props = {
...
 	className,
 	features,
 	isExpanded: isExpandedByDefault,
+	productSlug,
 } ) => {
+	const trackProps = productSlug ? { product_slug: productSlug } : {};
 	const trackShowFeatures = useTrackCallback(
 		undefined,
-		'calypso_jetpack_show_product_card_features'
+		'calypso_product_features_open',
```
Comment (rcanepa) [inline client/components/jetpack/card/jetpack-product-card/features.tsx:40]: Renamed it to make it follow the same naming convention the rest of the tracks use.

**Comment by rcanepa** [comment]
> @rcanepa , when I select 1 product, it appears that the `calypso_cart_product_add` event is firing twice, (but I only have one item in the cart), (in this case, Jetpack Anti Spam) - See screenshot:
> Screenshot on 2020-09-30 at 10-08-44
> 
> When I select Backup Daily and choose the upsell to add Scan, it appears that Backup is being added to cart twice (like in my example above), but Scan is being added to cart once. See screenshot:
> Screenshot on 2020-09-30 at 10-10-34
> 
> Is this the case that `calypso_cart_product_add` is firing twice for the first product added to the cart? Or maybe I'm interpreting the console debug events wrong?
> 
> Can you duplicate?

Yes @elliottprogrammer. I can definitely see the duplicates. But there is subtle difference between events that appear to be same. See this example:

image

One of them starts with `Record` and the other one with `Recording`. The first says `called with props` and the latter `with actual props`. 

So, I think this is because events are not immediately registered, they are put in a queue first. I'm not 100% sure about this, so I'll try to figure out what's going on.

**Comment by rcanepa** [comment]
@elliottprogrammer I have an answer for the duplicate records. Here https://github.com/Automattic/wp-calypso/blob/master/packages/calypso-analytics/src/tracks.ts#L214 you can see when those console statements are logged. The first of them happens before two checks (one of them can prevent the event from being tracked), and the second one is when the event is being tracked. This is to say that it's fine to see two messages per event.

**Comment by elliottprogrammer** [comment]
> > @rcanepa , when I select 1 product, it appears that the `calypso_cart_product_add` event is firing twice, (but I only have one item in the cart), (in this case, Jetpack Anti Spam) - See screenshot:
> > Screenshot on 2020-09-30 at 10-08-44
> > When I select Backup Daily and choose the upsell to add Scan, it appears that Backup is being added to cart twice (like in my example above), but Scan is being added to cart once. See screenshot:
> > Screenshot on 2020-09-30 at 10-10-34
> > Is this the case that `calypso_cart_product_add` is firing twice for the first product added to the cart? Or maybe I'm interpreting the console debug events wrong?
> > Can you duplicate?
> 
> Yes @elliottprogrammer. I can definitely see the duplicates. But there is subtle difference between events that appear to be same. See this example:
> 
> image
> 
> One of them starts with `Record` and the other one with `Recording`. The first says `called with props` and the latter `with actual props`.
> 
> So, I think this is because events are not immediately registered, they are put in a queue first. I'm not 100% sure about this, so I'll try to figure out what's going on.

@rcanepa , yes I see that one starts with `Record, called with props` and the other one with `Recording, called with actual props`, so when a product is added to the cart, I would expect to see these 2 console debug events for one product added to the cart.  But I'm seeing 4 of these debug events for 1 product added to cart.  See my first screenshot where I show 4 console debug statements.

**Comment by rcanepa** [comment]
> @rcanepa , yes I see that one starts with `Record, called with props` and the other one with `Recording, called with actual props`, so when a product is added to the cart, I would expect to see these 2 console debug events for one product added to the cart. But I'm seeing 4 of these debug events for 1 product added to cart. See my first screenshot where I show 4 console debug statements.

I understand now. Thanks for catching that. I'll work on a fix immediately.

**Comment by rcanepa** [comment]
@elliottprogrammer the problem was that `calypso_cart_product_add` is being used by the checkout code to track when a product has been added to the cart. To fix this, I renamed our event to `calypso_product_checkout_click` which means the intent of the user to go to the checkout page with the product. 

It's important to understand that in some flows (e.g. Jetpack Connect), the user may need to select a site and log in before arriving to the checkout page. This means that the user might decide to abort the process before we even add the product to the cart. In short, `calypso_cart_product_add` wasn't the best name for the event.

**Review by elliottprogrammer — APPROVED** [review]
Tested all the events on all 3 sections (Calypso Plans page, Jetpack Connect Plans page, and Calypso Green Plans page)
All events and page views are dispatching correctly.
Code looks fantastic. works great.
Nice work Renzo! 🙂
LGTM! 👍

**Comment by rcanepa** [comment]
Just pushed a new commit that adds the `calypso_product_jpfree_click` track. This will get dispatched when users clicks on the Jetpack Free card CTA.

**Pull request merged**